libafl/src/events/mod.rs

```diff
@@ -436,130 +419,208 @@ pub trait EventFirer: UsesState {
 where
-    Self::State: HasMetadata + HasExecutions + HasLastReportTime,
+    EM: AdaptiveSerializer,
+    OT: MatchNameRef + Serialize,
 {
-    /// Given the last time, if `monitor_timeout` seconds passed, send off an info/monitor/heartbeat message to the broker.
-    /// Returns the new `last` time (so the old one, unless `monitor_timeout` time has passed and monitor have been sent)
-    /// Will return an [`Error`], if the stats could not be sent.
-    fn maybe_report_progress(
-        &mut self,
-        state: &mut Self::State,
-        monitor_timeout: Duration,
-    ) -> Result<(), Error> {
-        let Some(last_report_time) = state.last_report_time() else {
-            // this is the first time we execute, no need to report progress just yet.
-            *state.last_report_time_mut() = Some(current_time());
-            return Ok(());
-        };
-        let cur = current_time();
-        // default to 0 here to avoid crashes on clock skew
-        if cur.checked_sub(*last_report_time).unwrap_or_default() > monitor_timeout {
-            // report_progress sets a new `last_report_time` internally.
-            self.report_progress(state)?;
+    match manager.time_ref() {
+        Some(t) => {
+            let exec_time = observers
+                .get(t)
+                .map(|o| o.last_runtime().unwrap_or(Duration::ZERO))
+                .unwrap();
+
+            let mut must_ser = (manager.serialization_time() + manager.deserialization_time())
+                * time_factor
+                < exec_time;
+            if must_ser {
+                *manager.should_serialize_cnt_mut() += 1;
+            }
+
+            if manager.serializations_cnt() > 32 {
+                must_ser = (manager.should_serialize_cnt() * 100 / manager.serializations_cnt())
+                    > percentage_threshold;
+            }
+
+            if manager.serialization_time() == Duration::ZERO
+                || must_ser
+                || manager.serializations_cnt().trailing_zeros() >= 8
+            {
+                let start = current_time();
+                let ser = postcard::to_allocvec(observers)?;
+                *manager.serialization_time_mut() = current_time() - start;
+
+                *manager.serializations_cnt_mut() += 1;
+                Ok(Some(ser))
+            } else {
+                *manager.serializations_cnt_mut() += 1;
+                Ok(None)
+            }
         }
-        Ok(())
+        None => Ok(None),
     }
+}
 
-    /// Send off an info/monitor/heartbeat message to the broker.
-    /// Will return an [`Error`], if the stats could not be sent.
-    fn report_progress(&mut self, state: &mut Self::State) -> Result<(), Error> {
-        let executions = *state.executions();
-        let cur = current_time();
+/// Default implementation of [`ProgressReporter::maybe_report_progress`] for implementors with the
+/// given constraints
+pub fn default_maybe_report_progress<PR, S>(
```

Comment:
shouldn't we do `std` to follow other names in the codebase?

Reply:
i'll replace all at last

Reply:
Also, we should document in the trait if a std method is available